### PR TITLE
1581 `rake policy:load` caused start script to fail

### PR DESCRIPTION
**Pull request by h-artzi**
### What does this PR do?
- _What's changed? Why were these changes made?_
Configures the start script to restart Conjur when authenticator flags are used. This was done in order to have a clean startup script without the need to authenticate again
In addition, it addresses the issues regarding module/class visibility that the conjurctl calls were running into
- _How should the reviewer approach this PR, especially if manual tests are required?_
run `./dev/start.sh --authn-ldap` and make sure there are no errors
- _Are there relevant screenshots you can add to the PR description?_
```
.....
+ echo 'conjur server is up and ready'
conjur server is up and ready
+ return 0
+ echo 'Creating user alice'
Creating user alice
+ docker-compose exec client conjur policy load root /src/conjur-server/dev/files/policy.yml
Loaded policy 'root'
{
  "created_roles": {
  },
  "version": 2
}
+ echo 'killing the conjur server process'
killing the conjur server process
+ docker-compose exec conjur /src/conjur-server/dev/files/killConjurServer
++ docker-compose ps -q conjur
+ docker exec -e CONJUR_AUTHN_API_KEY=322d9n935wa3gh3v1dt6z16edfte3gvpdq91sbtwj8wtdx6n305mtyh -e LDAP_URI=ldap://ldap-server:389 -e LDAP_BASE=dc=conjur,dc=net -e 'LDAP_FILTER=(uid=%s)' -e LDAP_BINDDN=cn=admin,dc=conjur,dc=net -e LDAP_BINDPW=ldapsecret -e CONJUR_AUTHENTICATORS=authn,authn-config/env,authn-ldap/test,authn-ldap/secure -it --detach-keys 'ctrl-\' c58035e72bedc327681e248346ba3589a28340048cc19ab21fa9a8284894ae29 bash
root@c58035e72bed:/src/conjur-server# exit
```

### What ticket does this PR close?
Connected to #1581 

#### Change log
- [ ] The CHANGELOG has been updated, or
- [x] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [x] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] This PR does not require updating any documentation, or
- [x] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs

### CODE REVIEW

File: lib/root_loader.rb [hunk 1]
```diff
@@ -1,6 +1,10 @@
 # frozen_string_literal: true
 
+puts("HELP")
+puts($LOAD_PATH)
```
Comment (izgeri) [inline lib/root_loader.rb:4]: should be removed before merge

**Comment by izgeri** [comment]
@h-artzi this PR does more than what's described above, right? it also fixes an issue I saw where I ran Conjur in a docker-compose env and set `CONJUR_ADMIN_PASSWORD` in the container, and when it ran the rake task (? not sure which one) on startup the container died on:
```
Setting 'admin' password
rake aborted!
NameError: uninitialized constant Loader::ReplacePolicy
/opt/conjur-server/lib/root_loader.rb:62:in `accepted_roles'
/opt/conjur-server/lib/root_loader.rb:32:in `block in load'
/var/lib/gems/2.5.0/gems/sequel-4.49.0/lib/sequel/database/transactions.rb:224:in `_transaction'
/var/lib/gems/2.5.0/gems/sequel-4.49.0/lib/sequel/database/transactions.rb:199:in `block in transaction'
/var/lib/gems/2.5.0/gems/sequel-4.49.0/lib/sequel/database/connecting.rb:301:in `block in synchronize'
/var/lib/gems/2.5.0/gems/sequel-4.49.0/lib/sequel/connection_pool/threaded.rb:107:in `hold'
/var/lib/gems/2.5.0/gems/sequel-4.49.0/lib/sequel/database/connecting.rb:301:in `synchronize'
/var/lib/gems/2.5.0/gems/sequel-4.49.0/lib/sequel/database/transactions.rb:165:in `transaction'
/opt/conjur-server/lib/root_loader.rb:17:in `load'
/opt/conjur-server/lib/tasks/policy.rake:51:in `block (2 levels) in <top (required)>'
/var/lib/gems/2.5.0/gems/rake-13.0.1/exe/rake:27:in `<top (required)>'
Tasks: TOP => policy:load
(See full trace by running task with --trace)
error: exit
```

I think it would be good to capture that in #1581 (which as of this writing still says "Temp") and in the change log. but fwiw I did test these changes, and they do fix my issue!!

File: dev/start [hunk 2]
```diff
@@ -256,6 +257,24 @@ env_args="$env_args -e CONJUR_AUTHENTICATORS=$enabled_authenticators"
 
 docker-compose up -d --no-deps $services
 
+if [[ $ENABLE_AUDIT = true ]]; then
```
Comment (micahlee) [inline dev/start:260]: This block is a duplicate of https://github.com/cyberark/conjur/blob/de82a4a4b594a9cca62a0f475eb5f3f8cacc1c15/dev/start#L120-L127

Is this intentional, or the result of a bad rebase?

**Comment by micahlee** [comment]
I agree that we should add a CI test that uses `conjurctl policy load` as a smoke test for this. We're in the midst of continued refactoring that does interact with policy loading, and until we're actually ready to nix the conjurctl policy stuff for good, we should put a safety net under it to mitigate the chance of this occurring again.

**Comment by micahlee** [comment]
I also agree that we should clean up this issue and PR from the temp state it's currently in.

File: dev/start (hunk 2 above)
Comment (h-artzi) [reply dev/start]: bad rebase

**Comment by h-artzi** [comment]
Failed build prior to fix: https://jenkins.conjur.net/job/cyberark--conjur/job/1581-conjurctl-start/10/execution/node/153/log/

**Comment by h-artzi** [comment]
The issue that we were running into was not an error only occurred when the rake task was being used.

File: cucumber/policy/features/support/policy.yml [hunk 3]
```diff
@@ -0,0 +1 @@
+- !user test
```
Comment (orenbm) [inline cucumber/policy/features/support/policy.yml:1]: missing new line

File: cucumber/policy/features/rake.feature [hunk 4]
```diff
@@ -0,0 +1,10 @@
+Feature: Rake task to load Conjur policy
+
+Conjur includes a Rake task (`rake policy:load`) for loading policies from 
+within the Conjur container. This rake task is used by the `conjurctl policy 
+load
+
+  Scenario: Load a simple policy using `rake policy:load`
+    When I load a policy from file "policy.yml" using conjurctl
+    Then user "test" exists
+
```
Comment (orenbm) [inline cucumber/policy/features/rake.feature:10]: missing new line

**Comment by orenbm** [comment]
Hey @h-artzi ! I'm reviewing your PR and don't really understand how it is related to the topic. The only thing you changed in the `start` script is to add a `default_authenticators` variable but it doesn't seem to have any impact. Was there really an issue with the `start` script that you fixed? If not, can you rename the PR and update the description to better fit the content? Same goes for the commit "always wait for conjur".

Also - It seems that the commit "add necessary require statement" can also be broken as it does more than what it claims to do. You can move the part where you move classes into their own file into another commit.

These steps will make the review process much easier. Thanks :)

**Comment by h-artzi** [comment]
The issue with the start script was the need to authenticate. However, the start file Oren checked earlier was outdated and I forgot to update it.
Before:
```
....
Creating dev_ldap-server_1 ... done
Recreating dev_conjur_1    ... done
Recreating dev_client_1    ... done
+ echo 'Creating user alice'
Creating user alice
+ docker-compose exec client conjur policy load root /src/conjur-server/dev/files/policy.yml
Enter your username to log into Conjur:
```
After:
```
.....
Creating user alice
+ docker-compose exec client conjur policy load root /src/conjur-server/dev/files/policy.yml
Loaded policy 'root'
{
  "created_roles": {
  },
  "version": 2
}
+ echo 'killing the conjur server process'
killing the conjur server process
+ docker-compose exec conjur /src/conjur-server/dev/files/killConjurServer
++ docker-compose ps -q conjur
+ docker exec -e CONJUR_AUTHN_API_KEY=29f35qq6tpr2ma727zhsmv3ga3v6y48d2kegh91324barb1s2a8z7 -e LDAP_URI=ldap://ldap-server:389 -e LDAP_BASE=dc=conjur,dc=net -e 'LDAP_FILTER=(uid=%s)' -e LDAP_BINDDN=cn=admin,dc=conjur,dc=net -e LDAP_BINDPW=ldapsecret -e CONJUR_AUTHENTICATORS=authn,authn-config/env,authn-ldap/test,authn-ldap/secure -it --detach-keys 'ctrl-\' 2d8571416199cb6b3ba2291f4fe5412446c2711925d3b3449537de01df6ae42d bash
root@2d8571416199:/src/conjur-server#
```

File: cucumber/policy/features/rake.feature (hunk 4 above)
Comment (orenbm) [reply cucumber/policy/features/rake.feature:11]: looks like the new line is still missing. Maybe you have a tab?

File: cucumber/policy/features/rake.feature [hunk 5]
```diff
@@ -0,0 +1,11 @@
+Feature: Rake task to load Conjur policy
+
+Conjur includes a Rake task (`rake policy:load`) for loading policies from 
+within the Conjur container. This rake task is used by the `conjurctl policy 
+load
```
Comment (orenbm) [inline cucumber/policy/features/rake.feature:5]: load`

File: cucumber/policy/features/step_definitions/rake_steps.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+When(/^I load a policy from file "([^"]*)" using conjurctl/) do |filename|
+  absolute_path = "#{File.dirname __FILE__}/../support/#{filename}"
+  rake_task = ["rake", "policy:load[cucumber,#{absolute_path}]"]
```
Comment (orenbm) [inline cucumber/policy/features/step_definitions/rake_steps.rb:5]: please add a space after the last comma

File: dev/start [hunk 7]
```diff
@@ -256,6 +257,15 @@ env_args="$env_args -e CONJUR_AUTHENTICATORS=$enabled_authenticators"
 
 docker-compose up -d --no-deps $services
 
+# If the enabled authenticators changed after initial Conjur configuration,
+# then docker-compose recreates the container to set the environment variable,
+# and we need to restart the Conjur server process.
```
Comment (orenbm) [inline dev/start:262]: Nice catch! In which scenario did you see this? Does it really matter that we run the server again before loading the policy with alice? It seems that it's the only thing happening before killing the process again.

File: lib/root_loader.rb [hunk 8]
```diff
@@ -1,6 +1,7 @@
 # frozen_string_literal: true
 
 require 'json'
+require 'logs'
```
Comment (orenbm) [inline lib/root_loader.rb:4]: So what happens if this is missing? Why did we need to add it?

**Comment by orenbm** [comment]
Thanks @h-artzi , it looks much better!

Can you please read [this guide](https://chris.beams.io/posts/git-commit/) and improve your commits accordingly? They will make them better. I also prefer the title `Test Rake task to load Conjur policy` rather than `test rake policy:load` as it's easier to read. But that's a small nit :)

File: dev/start (hunk 7 above)
Comment (h-artzi) [reply dev/start:262]: Micah actually caught it and when you run `./start --authn-ldap` for example. If we don't run the service again the user needs to manually authenticate in order to add the policy with Alice. The difference between the two runs is highlighted [here](https://github.com/cyberark/conjur/pull/1582#issuecomment-638229844)

File: lib/root_loader.rb (hunk 8 above)
Comment (h-artzi) [reply lib/root_loader.rb:4]: It gives an error regarding not being able to find the following class and hence fails. 
```
PrimarySchema = Util::TrackableErrorClass.new(
        msg:  "Primary schema is {0-primary-schema}",
        code: "CONJ00034I"
      )
``` 
found in [logs.rb](https://github.com/cyberark/conjur/blob/f901b847da8cdc10b659d71167cba5f406b1eb77/app/domain/logs.rb#L13)
If you would like I can run the script and copy and paste the logs

File: dev/start (hunk 7 above)
Comment (orenbm) [reply dev/start:262]: Oh nice! I remember seeing this behaviour once. I think it happens only with `authn-ldap`. thanks!

File: lib/root_loader.rb (hunk 8 above)
Comment (orenbm) [reply lib/root_loader.rb:4]: ah. that's code that i added recently. weird that it doesn't fail the build in jenkins. Thanks for the explanation.

File: cucumber/policy/features/rake.feature (hunk 4 above)
Comment (orenbm) [reply cucumber/policy/features/rake.feature:11]: The new line is missing still. my guess is that you have a tab

File: dev/start [hunk 9]
```diff
@@ -256,6 +257,16 @@ env_args="$env_args -e CONJUR_AUTHENTICATORS=$enabled_authenticators"
 
 docker-compose up -d --no-deps $services
 
+# Hack: this file can be refactored to avoid this
```
Comment (orenbm) [inline dev/start:260]: Is this line here on purpose?

**Pull request merged**